### PR TITLE
feat(Rules): RHICOMPL-1403 implement associate rules mutation

### PR DESCRIPTION
**Pull request by skateman**
Implemented the `AssociateRules` mutation inspired by `AssociateSystems` analogously to the REST endpoint for updating rules under a profile.

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/graphql/mutations/profile/associate_rules.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+      argument :id, ID, required: true
+      argument :rule_ids, [ID], required: true
+      field :profile, Types::Profile, null: true
+
+      def resolve(args = {})
+        profile = find_profile(args[:id])
+        if profile
+          profile.update_rules(ids: args[:rule_ids])
+          audit_mutation(profile)
```
Comment (vkrizan) [inline app/graphql/mutations/profile/associate_rules.rb:17]: `update_rules` returns number of added and removed rules. Could it be used as a condition for the audit logging? The numbers might be included within the message.

File: app/graphql/mutations/profile/associate_rules.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        end
+        { profile: profile }
+      end
+
+      include ProfileHelper
```
Comment (vkrizan) [inline app/graphql/mutations/profile/associate_rules.rb:22]: Nit: Could be moved to the top.

File: app/graphql/types/mutation.rb [hunk 3]
```diff
@@ -13,6 +13,7 @@ class Mutation < Types::BaseObject
     field :deleteTestResults, mutation: Mutations::TestResult::Delete
     field :associateSystems, mutation: Mutations::Profile::AssociateSystems
     field :associateProfiles, mutation: Mutations::System::AssociateProfiles
+    field :associateRules, mutation: Mutations::Profile::AssociateRules
```
Comment (vkrizan) [inline app/graphql/types/mutation.rb:16]: New types and mutations need the schema file to be updated. See test failure and/or run `bundle exec rake dump_schema`.

**Review by vkrizan — COMMENTED** [review]
Looks good. Last thing is to update the GraphQL schema (see inline).

File: app/graphql/mutations/profile/associate_rules.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+module Mutations
+  module Profile
+    # Mutation to associate systems with a profile
+    class AssociateRules < BaseMutation
+      graphql_name 'associateRules'
+
+      argument :id, ID, required: true
+      argument :rule_ids, [ID], required: true
```
Comment (vkrizan) [inline app/graphql/mutations/profile/associate_rules.rb:10]: @bastilian Would frontend send UUIDs or does it use ref_ids instead?

File: app/graphql/mutations/profile/associate_rules.rb (hunk 1 above)
Comment (skateman) [reply app/graphql/mutations/profile/associate_rules.rb:17]: Just making sure that I understand :thinking: you want me to emit an audit message like this:
```
Updated N rule associations of profile #{profile.id}
```
?

Comment (vkrizan) [reply app/graphql/mutations/profile/associate_rules.rb:17]: Yes. Check out profiles controller for inspiration. It logs smth similar.

File: app/graphql/mutations/profile/associate_rules.rb (hunk 4 above)
Comment (bastilian) [reply app/graphql/mutations/profile/associate_rules.rb:12]: @vkrizan @skateman The Frontend does send ref_ids at the moment. I'm not opposed to changing that at all btw., but maybe later on, or if we encounter issues with using the ref_id.

Comment (vkrizan) [reply app/graphql/mutations/profile/associate_rules.rb:12]: @skateman @bastilian What do you guys suggest then? Can we use UUID ids to keep the APIs consistent and update the frontend to do the use them?

Comment (akofink) [reply app/graphql/mutations/profile/associate_rules.rb:12]: I recommend using UUIDs like the REST API if using `rule_ids`. If we want to support ref_ids in addition, we should add another argument `rule_ref_ids`.

**Review by akofink — APPROVED** [review]
Works as expected. Thanks, @skateman!

NB: Sending an empty rules array assigns the canonical set of rules, the same way the REST API works. This is how `Profile#update_rules` works. I'm not sure any user would want to set 0 rules on a profile, but it could come up again. In any case, this PR doesn't alter that behavior; if we want to change it, we should submit another jira and fix it separately.

File: test/graphql/mutations/associate_rules_test.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class AssociateRulesMutationTest < ActiveSupport::TestCase
+  setup do
+    profiles(:one).update account: accounts(:one),
+                          policy: policies(:one)
+    users(:test).update account: accounts(:one)
```
[the remark this answers is not in the document]
Comment (skateman) [reply test/graphql/mutations/associate_rules_test.rb:9]:
```
No fixture named 'one' found for fixture set 'users'
```
:shrug:

[the remark this answers is not in the document]
Comment (akofink) [reply test/graphql/mutations/associate_rules_test.rb:9]: Oops, sorry about that. I was thinking of the accounts fixture. :+1:

File: app/graphql/mutations/profile/associate_rules.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      def resolve(args = {})
+        profile = find_profile(args[:id])
+        if profile
+          rules_added, rules_removed = profile.update_rules(
+            ids: args[:rule_ids]
+          )
+          audit_mutation(profile, rules_added, rules_removed)
+        end
+        { profile: profile }
+      end
```
Comment (akofink) [inline app/graphql/mutations/profile/associate_rules.rb:24]: Recommend wrapping this in a transaction

Comment (akofink) [reply app/graphql/mutations/profile/associate_rules.rb:24]: It can be done in a followup PR.

Comment (skateman) [reply app/graphql/mutations/profile/associate_rules.rb:24]: @akofink does this mean that we should have a transaction around the [associate hosts](https://github.com/RedHatInsights/compliance-backend/blob/master/app/graphql/mutations/profile/associate_systems.rb#L16) as well?

Comment (akofink) [reply app/graphql/mutations/profile/associate_rules.rb:24]: It can't hurt! Thanks for noticing that.

**Pull request merged**